### ISSUE TITLE
TLSv1.3 unexpected InfoCallback after handshake completed

### ISSUE THREAD

**Issue by magsys** [openssl/openssl#5721]
Our Delphi SSL library uses info_callback to check handshake status, in particlar to detect renegotiaton for older TLS protocols.  

With TLSv1.3, once the initial handshake is complete,  there is a second ICB SSL_handshake_start event, which logging shows precedes a New Session Ticket packet.  This ICB breaks our code, because we closed the connection to stop renegotiaton.   Fixed by ignoring it for TLSv1.3, but should it be called atall? 

11:36:25:844 04326370 Start Ssl ReadBIO
11:36:25:856 04326370 ICB> SSL_handshake_start: where=00000010, state=before SSL initialization
11:36:25:864 04326370 ICB> SSL_CB_HANDSHAKE_START
lots cut
11:36:26:256 04326370 ICB> SSL_handshake_done: where=00000020, state=SSL negotiation finished successfully
11:36:26:264 04326370 ICB> SSL_CB_HANDSHAKE_DONE, Error ok
11:36:26:274 04326370 SslHandshakeDone(0) Error=916. Secure connection with TLSv1.3, cipher TLS_AES_256_GCM_SHA384, 256 secret bits (256 total), session reused=False
more cut
11:36:26:345 ProtoMsg: TLS 1.3, State: SSL negotiation finished successfully, Recv, DataLen: 1, Data= 16
!!! unexpected ICB follows
11:36:26:356 04326370 ICB> SSL_handshake_start: where=00000010, state=SSL negotiation finished successfully
11:36:26:362 ProtoMsg: TLS 1.3 Handshake: New Session Ticket, State: SSLv3/TLS read server session ticket, Recv, DataLen: 274, Data=xx 
11:36:26:372 ProtoMsg: None, State: SSL negotiation finished successfully, Recv, DataLen: 5, Data= 17030319EA
11:36:26:396 ProtoMsg: TLS 1.3, State: SSL negotiation finished successfully, Recv, DataLen: 1, Data= 17

Angus

**Comment by mattcaswell** [comment]
I think this is correct behaviour. This is effectively the start of a new exchange of handshake messages (well one message to be precise). The original handshake has completed and the NewSessionTicket message is specified in the spec with a record type of "handshake" so it seems reasonable to me that its arrival should be signalled through an SSL_CB_HANDSHAKE_START event.

**Comment by magsys** [comment]
If so, should there also be a balancing SSL_CB_HANDSHAKE_DONE event?  

Perhaps this should be a new  SSL_CB_POST_HANDSHAKE_START event?  Makes it more backward compatible?  

Angus

**Comment by mattcaswell** [comment]
> If so, should there also be a balancing SSL_CB_HANDSHAKE_DONE event?

Probably, yes.

> Perhaps this should be a new SSL_CB_POST_HANDSHAKE_START event? Makes it more backward compatible?

Yes, that could work.

**Comment by mattcaswell** [comment]
Fix for the missing handshake done events in #5874. I've also updated the documentation to explain the meaning of the various events in a TLSv1.3 context. I opted not to introduce a new post handshake start event. While it might have helped you specifically to avoid this issue I'm not sure that would be the case in general. Ultimately TLSv1.3 works differently to TLSv1.2 so it is not surprising that the info callbacks are going to be a bit different. There's not much we can do about that if an application is not expecting it.
